src/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.CoreRT.cs

```diff
@@ -0,0 +1,71 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    public partial struct GCHandle
+    {
+        private IntPtr InternalAlloc(object value, GCHandleType type)
+        {
+            if (type == GCHandleType.Pinned)
+            {
+                GCHandleValidatePinnedObject(value);
+            }
+
+            return RuntimeImports.RhHandleAlloc(value, type);
+        }
+
+        private void InternalFree(IntPtr handle) => RuntimeImports.RhHandleFree(handle);
+
+        private object InternalGet(IntPtr handle) => RuntimeImports.RhHandleGet(handle);
+
+        private void InternalSet(IntPtr handle, object value, bool isPinned)
+        {
+            if (isPinned)
+            {
+                GCHandleValidatePinnedObject(value);
+            }
+
+            RuntimeImports.RhHandleSet(handle, value);
+        }
+
+        private unsafe IntPtr InternalAddrOfPinnedObject(IntPtr handle)
+        {
+            // Get the address of the pinned object.
+            // The layout of String and Array is different from Object
+
+            object target = InternalGet(handle);
+
+            if (target == null)
+            {
+                return default(IntPtr);
+            }
+
+            if (target is string targetAsString)
+            {
+                return (IntPtr)Unsafe.AsPointer(ref targetAsString.GetRawStringData());
+            }
+
+            if (target is Array targetAsArray)
+            {
+                return (IntPtr)Unsafe.AsPointer(ref targetAsArray.GetRawArrayData());
+            }
+
+            return (IntPtr)Unsafe.AsPointer(ref target.GetRawData());
+        }
+
+        private static void GCHandleValidatePinnedObject(object obj)
```

Comment:
This can be in shared too if we wanted to ... there is `Marshal.IsPinnable` that does this check.

Reply:
We can just change the call sites in this file then to use Marshal.IsPinnable then... this method isn't used by the shared GCHandle.cs, only by the other methods here.

Reply:
I meant that all the error checking can be in shared. The error checking and need to throw exceptions from the FCall makes the CreateHandle FCall much slower than necessary because of it needs to erect `HELPER_METHOD_FRAME`. If all error checkin is moved to C# and the FCall returns NULL on failure like in CoreRT, it should be faster than what it is today.

Reply:
Ah.  Ok, I'll try that.

Reply:
(It is fine to do this in follow up change if it looks too hard.)

Reply:
Seems straightforward.

Reply:
@jkotas, so if in coreclr's GCHandleInternalSet I remove the call to ValidatePinnedObject, can I then also remove the HELPER_METHOD_FRAME_BEGIN_1 / HELPER_METHOD_FRAME_END macros?

Reply:
I addressed this in the coreclr changes, but I'm not entirely sure about them.